keps/sig-apps/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,907 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+A StatefulSet of `N` replicas implicitly numbers pods from ordinal `0` to `N-1`. The end ordinal (`N-1`) can be controlled with the `replicas` field. The goal of this feature is to allow a StatefulSet’s first ordinal to start from any natural number `k`. This enables StatefulSet ordinals from `k` to `N+k-1`
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+This feature is motivated by the use case of migrating a StatefulSet across a namespace or a Kubernetes cluster. Existing approaches to this problem include:
+
+1. Back and restore: This approach takes a backup of an application (StatefulSet, underlying storage), and re-creates it in a different location. This introduces application downtime, the duration of time between old StatefulSet termination and new StatefulSet recreation.
+2. Pod level migration: Using `--cascade=orphan` when deleting a StatefulSet preserves the pods. This allows an application operator to evict and reschedule pods individually. However, as pods are ephemeral, this requires the application operator to emulate the behavior of the StatefulSet, to reschedule pods as they restart, or are evicted and rescheduled.
+
+Migrating a StatefulSet in slices allows for gradual migration of the application, as liveness is maintained. Consider the scenario of transferring pod ordinal ownership from an old StatefulSet with `N` pods to a new StatefulSet with `0` pods. Further, to maintain application availability, no more than `d` pods should be unavailable at any time during the transfer.
+
+StatefulSets are implicitly numbered starting at ordinal `0`. When pods are being deployed, they are created [sequentially in order](StatefulSet) from pod `0` to pod `N-1`. When pods are being deleted, they are terminated in reverse order from pod `N-1` to pod `0`. This behavior limits the migration scenario where an application operator wants to scale down pods in the old StatefulSet and scale up pods in the new StatefulSet. If pod `N-1` is removed from the old StatefulSet, there is no mechanism to create only pod `N-1` in a new StatefulSet without creating pods `[0, N-2]` as well. To do so would lead to the presence of duplicate pod ordinals (eg: pod `0` would exist in both StatefulSets).
```

Comment:
I wonder what is preventing app admin from migrating pod i  to become pod N-i in the new cluster? Does the order matter to the user?

Reply:
I don't understand either. Maybe the DNS domain can be the same, which is `$(pod-name).$(service-name).$(namespace).svc.cluster.local`, but that should be meaningless since it is in a different cluster.

Reply:
Depending on the application and configuration, the ordinal that is assigned to a pod can be significant (eg: if the pod name is passed down using a download API to the underlying application, or the application uses the hostname to extract the ordinal). This is one reason to preserve the ordinal "i" when migrating a pod from the old to the new cluster.

Additionally if the StatefulSet is using a volumeClaimTemplate, the PV/PVC will be linked to the pod through ordinal "i". If underlying data is to be transferred or referenced in a new cluster, this adds additional complexity for the app admin. Consider migrating a pod with ordinal "m"  remapped to pod "n" in a new cluster: the PV/PVC name in the new cluster would need to be updated to reference ordinal "n".